From dd710281580135b40c98e95bc6a9ad562052f010 Mon Sep 17 00:00:00 2001
From: Dudemanguy <random342@airmail.cc>
Date: Wed, 22 Jan 2025 15:50:12 -0600
Subject: [PATCH] opengl/context: require swap_buffers param for FenceSync

Commits 8854c74 and 5ae0e0f eliminated the external swapchain API that
the render backend (vo_libmpv) was previously using. However, it was
missed that this was being used to skip the gl->FenceSync calls and that
logic change was not properly accounted for. This meant that vo_libmpv
would leak fences since ra_gl_ctx_swap_buffers is never called to clean
them up.

Fix this by making sure the ra_ctx_params have swap_buffers defined
before calling gl->FenceSync. All platforms are required to implement
this anyway with the exception of vo_libmpv since it's a funny special
case. Fixes #17217.
---
 video/out/opengl/context.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/video/out/opengl/context.c b/video/out/opengl/context.c
index df8a68ac66..3f8ff0385e 100644
--- a/video/out/opengl/context.c
+++ b/video/out/opengl/context.c
@@ -199,7 +199,7 @@ void ra_gl_ctx_swap_buffers(struct ra_ctx *ctx)
     struct priv *p = ctx->priv;
     GL *gl = p->gl;
 
-    if (gl->FenceSync) {
+    if (gl->FenceSync && ctx->ra_ctx->params.swap_buffers) {
         GLsync fence = gl->FenceSync(GL_SYNC_GPU_COMMANDS_COMPLETE, 0);
         if (fence)
             MP_TARRAY_APPEND(p, p->vsync_fences, p->num_vsync_fences, fence);